### PR TITLE
Quda interface

### PR DESCRIPTION
**Pull request by m-schroeck**
This is an interface for the QUDA library. So far only the "invert" executable supports inversions with Quda. Extension to the determinant monomials for the HMC is straightforward.

See the documentation (tex) for details!

### CODE REVIEW

**Comment by urbach** [comment]
Hi Mario, could you please merge 'master' into this branch. Thanks!

File: config.h.in [hunk 1]
```diff
@@ -1,34 +1,135 @@
 /* config.h.in.  Generated from configure.in by autoheader.  */
-#ifndef _CONFIG_H
```
Comment (urbach) [inline config.h.in]: hmm, why was this removed?

File: doc/main.bbl [hunk 2]
```diff
@@ -0,0 +1,195 @@
+\begin{thebibliography}{10}
```
Comment (urbach) [inline doc/main.bbl]: we don't need main.bbl under git control. It is generated via `bibtex`

File: operator.c [hunk 3]
```diff
@@ -264,12 +278,20 @@ void op_invert(const int op_id, const int index_start, const int write_prop) {
...
 	  g_precWS=NULL;
 	}
-	
-	optr->iterations = invert_eo( optr->prop0, optr->prop1, optr->sr0, optr->sr1,
-				      optr->eps_sq, optr->maxiter,
-				      optr->solver, optr->rel_prec,
-				      0, optr->even_odd_flag,optr->no_extra_masses, optr->extra_masses, optr->solver_params, optr->id );
-	
+	if(optr->use_qudainverter) {
+#ifdef QUDA
```
Comment (urbach) [inline operator.c]: wouldn't this switch be better at a lower level? e.g. in `invert_eo.c`?

File: read_input.l [hunk 4]
```diff
@@ -727,6 +730,17 @@ inline void rmQuotes(char *str){
   }
 }
 
+<WILSONOP,TMOP,DBTMOP,CLOVEROP,DBCLOVEROP>{
```
Comment (urbach) [inline read_input.l]: how does this fit with the `BeginGPUInit` section in the input file?

**Comment by urbach** [comment]
line indentation is also different compare to the rest. We follow GNU with 4 spaces...

**Comment by urbach** [comment]
does this also work with a scalar build, i.e. without an MPI compiler?

File: quda_interface.c [hunk 5]
```diff
@@ -0,0 +1,786 @@
...
+				}
+
+	loadGaugeQuda((void*)gauge_quda, &gauge_param);
+}
+
+
+// reorder spinor to QUDA format
+void reorder_spinor_toQuda( double* spinor, QudaPrecision precision, int doublet, double* spinor2 )
+{
+	double startTime = MPI_Wtime();
```
Comment (urbach) [inline quda_interface.c]: use the custom `gettime` please, defined in `gettime.h` / `gettime.c`.

File: quda_interface.c [hunk 6]
```diff
@@ -0,0 +1,786 @@
...
+QudaInvertParam inv_param;
+
+// pointer to the QUDA gaugefield
+double *gauge_quda[4];
+
+// pointer to a temp. spinor, used for reordering etc.
+double *tempSpinor;
+
+// function that maps coordinates in the communication grid to MPI ranks
+int commsMap(const int *coords, void *fdata)
```
Comment (urbach) [inline quda_interface.c]: maybe this better becomes a generally available function?

File: quda_interface.c [hunk 7]
```diff
@@ -0,0 +1,786 @@
...
+{
+#if USE_LZ_LY_LX_T
+	int n[4] = {coords[3], coords[2], coords[1], coords[0]};
+#else
+	int n[4] = {coords[3], coords[0], coords[1], coords[2]};
+#endif
+
+	int rank;
+
+	MPI_Cart_rank( g_cart_grid, n, &rank );
```
Comment (urbach) [inline quda_interface.c]: bracket with `#ifdef MPI`

**Comment by urbach** [comment]
it seems I can at least compile without MPI compiler. Will it crash?

**Comment by urbach** [comment]
please don't use tabs... (even if still sometimes present in the code). E.g. emacs knows `untabbify` as a command...

File: doc/main.bbl (hunk 2 above)
Comment (m-schroeck) [inline doc/main.bbl]: removed.

File: config.h.in (hunk 1 above)
Comment (m-schroeck) [inline config.h.in]: I had to regenerate it with autoheader. Is there a problem? Should we revert it to the previous state?

File: read_input.l (hunk 4 above)
Comment (m-schroeck) [inline read_input.l]: The BeginGPUInit is from Florian Burger, so far I'm not using that. Do you want me to adapt it?

File: quda_interface.c (hunk 5 above)
Comment (m-schroeck) [inline quda_interface.c]: thanks for adding replacing the timing function etc. 
I removed the tabs in operator.c and quda_interface.c

File: quda_interface.c (hunk 6 above)
Comment (m-schroeck) [inline quda_interface.c]: right, could be useful elsewhere (e.g. in the qphix interface). where do you suggest do put it?

**Comment by m-schroeck** [comment]
Good point with the scalar build! There was in fact an issue (setting device has to be done explicitly in a scalar build). Fixed that. Just tested a scalar QUDA build linked to scalar tmLQCD (all with gcc, g++), compiles and runs fine! Added a note to the doc.

**Comment by urbach** [comment]
you mean you have to chose the device ID explicitly? This does not work on our cluster...

File: read_input.l (hunk 4 above)
Comment (urbach) [inline read_input.l]: hmm, what do you think? It offers more possibilities.

I think a combination of the two would be best. Currently this is completely independent, right?

File: config.h.in (hunk 1 above)
Comment (urbach) [inline config.h.in]: ah, okay, never used `autoheader`. We should definitly have the `#ifdef _CONFIG_H`. Otherwise, did you have to add so many new things to `config.h.in`? Do we need all the changes?

**Comment by kostrzewa** [comment]
I would like to pull this in now to prevent problems in the next few merges due to conflicts. For now we will simply live with the fact that in serial builds, device 0 is used by default.

**Comment by kostrzewa** [comment]
I'm a little bit hesitant because of the "default sloppy precision" in the operator, I don't think we want this in general, do we? I'd rather have the "less safe" version optional.

**Comment by kostrzewa** [comment]
Okay, I'll pull it in then, thanks for the clarification. I agree that mixed solvers with reliable updates are the way to go in general.
